Exercise8.java

```diff
@@ -0,0 +1,80 @@
+import java.util.Scanner;
+
+public class Exercise8{
+
+        public static void main(String args[]){
+                Scanner scanner = new Scanner(System.in);
+                String string = scanner.nextLine();
+                string = addSpacesIfMissing(string);
+                string = removeAllRedundantSpaces(string);
+                string = upperCaseAllCaptions(string);
+                System.out.println(string);
+        }
+
+        public static String removeAllRedundantSpaces(String str){
+                StringBuffer buffer = new StringBuffer();
+
+                for(String string: str.split("[\\s]"))
+                {
+                        if(!string.isEmpty()){
+                                if(string.equalsIgnoreCase("?") || string.equalsIgnoreCase("!")
+                                        || string.equalsIgnoreCase(",") || string.equalsIgnoreCase(".")){
+                                        buffer=buffer.append(string);
+                                } else {
+                                        buffer=buffer.append(" "+string);
+                                }
+                                
+                        }
+                }
+                return buffer.toString().trim();
+        }
+
+        public static String addSpacesIfMissing(String str){
+                String string = str;
+
+                for(int i=0; i<string.length()-1; i++){
+                        char c = string.charAt(i);
+                        if((c == '!' || c =='?' || c == '.' || c == ',') &&  c != ' '){
+                                string = string.substring(0, i+1).concat(" "+string.substring(i+1, string.length()));
```

Comment:
Câu lệnh
```java
string = string.substring(0, i+1).concat(" "+string.substring(i+1, string.length()));
```
là hơi khó hiểu.

Anh nên tạo một hàm để làm việc này.

Anh cũng nên tạo một đối tượng StringBuffer để làm việc và cái cách làm hiện tại là không tốt vì thao tác tạo chuỗi con, gộp chuỗi được thực hiện nhiều lần và trên toàn bộ chuỗi lớn. Ví dụ ở hàm trên là ok vì mỗi lần vòng for thực hiện, nó chỉ thêm 1 (hoặc 2) ký tự vào string buffer.